plugins/org.jkiss.dbeaver.ui.app.standalone/src/org/jkiss/dbeaver/ui/app/standalone/DBeaverApplication.java

```diff
@@ -310,50 +315,89 @@ private void markLocationReadOnly(Location instanceLoc) {
-
-        if (recentWorkspaces != null && recentWorkspaces.length > 1 && !ArrayUtils.contains(recentWorkspaces, WORKSPACE_DIR_CURRENT)) {
-            // Add default workspace in the recent list
-            boolean added = false;
-            for (int i = 0; i < recentWorkspaces.length; i++) {
-                if (recentWorkspaces[i] == null) {
-                    recentWorkspaces[i] = WORKSPACE_DIR_CURRENT;
-                    added = true;
-                    break;
-                }
-            }
-            if (!added) {
-                recentWorkspaces[recentWorkspaces.length - 1] = WORKSPACE_DIR_CURRENT;
-            }
-            launchData.setRecentWorkspaces(recentWorkspaces);
-            launchData.writePersistedData();
-        }
-
-        if (!ArrayUtils.isEmpty(recentWorkspaces)) {
-            String lastWorkspace = recentWorkspaces[0];
-            if (!CommonUtils.isEmpty(lastWorkspace) && !WORKSPACE_DIR_CURRENT.equals(lastWorkspace)) {
-                try {
-                    final URL selectedWorkspaceURL = new URL(
-                        "file",  //$NON-NLS-1$
-                        null,
-                        lastWorkspace);
-                    instanceLoc.set(selectedWorkspaceURL, true);
-
-                    return true;
-                } catch (Exception e) {
-                    System.err.println("Can't set IDE workspace to '" + lastWorkspace + "'");
-                    e.printStackTrace();
-                }
+        Collection<String> recentWorkspaces = getRecentWorkspaces(instanceLoc);
+        if (recentWorkspaces.isEmpty()) {
+            return false;
+        }
+        String lastWorkspace = recentWorkspaces.iterator().next();
+        if (!CommonUtils.isEmpty(lastWorkspace) && !WORKSPACE_DIR_CURRENT.equals(lastWorkspace)) {
+            try {
+                final URL selectedWorkspaceURL = new URL(
+                    "file",  //$NON-NLS-1$
+                    null,
+                    lastWorkspace);
+                instanceLoc.set(selectedWorkspaceURL, true);
+
+                return true;
+            } catch (Exception e) {
+                System.err.println("Can't set IDE workspace to '" + lastWorkspace + "'");
+                e.printStackTrace();
             }
         }
         return false;
     }
 
+    @NotNull
+    private static Collection<String> getRecentWorkspaces(@NotNull Location instanceLoc) {
+        ChooseWorkspaceData launchData = new ChooseWorkspaceData(instanceLoc.getDefault());
+        String[] arrayOfRecentWorkspaces = launchData.getRecentWorkspaces();
+        Collection<String> recentWorkspaces;
+        if (arrayOfRecentWorkspaces == null) {
+            recentWorkspaces = new ArrayList<>();
+        } else {
+            recentWorkspaces = Arrays.asList(arrayOfRecentWorkspaces);
+        }
+        int maxSize = recentWorkspaces.size();
+        recentWorkspaces.removeIf(Objects::isNull);
```

Comment:
This call will fail because `Arrays.asList` returns an unmodifiable list.

Reply:
Well, it's not modifiable, but `removeIf` will indeed fail. Thank you for noticing.

Also, the `removeIf` method of this particular `List` implementation works with O(n^2), which is not what I desire.